### ISSUE TITLE
[ENH] Contribute Distortion Index (Olteanu and al. 2019)

### ISSUE THREAD

**Issue by ceciledebezenac** [pysal/segregation#142]
Hello I am just opening this issue following a discussion with @darribas to contribute into segregation the statistic proposed in the following paper: 
[(Olteanu and al. 2019)](https://www.pnas.org/content/116/25/12250)
I have a working version in another repository[ here ](https://github.com/ceciledebezenac/segregation_index) I and I just wanted to kick off the discussion on how to integrate it in, and have your advice on the process.

**Comment by ljwolf** [comment]
Yes! Rad! This would be wonderful; I'm a [big fan of the paper](https://gist.github.com/ljwolf/b8b3f227b0eaf1975c8b2245ff403bcd) :grin:  

Taking a look at some small things that might help w/ maintenance, integration, & api consistency, we might want to make a few adjustments? It's real well documented & the code is very clean, so I think it should be simple to change a few things over. 

Not sure how @renanxcortes or @knaaptime feel, but I spot these things we'd wanna do for integration/consistency. Renan/Eli, you know better about the actual API of the functions, which we might want to also make consistent w/ the other measures. 

- [ ] **Use `scipy` more fully** 
  - [ ] [This distance could use `scipy.spatial.distance.pdist` or `sklearn.metrics.pairwise_distances`?](https://github.com/ceciledebezenac/segregation_index/blob/master/divergence/neighbourhood.py#L36), allowing us to use a more generic `metric='euclidean'` specification & use faster vectorized distance computations. 
  - [ ] Again should be similar for the [for dissimilarity calculation](https://github.com/ceciledebezenac/segregation_index/blob/master/divergence/neighbourhood.py#L59), which I think would apply in the `method` argument of `set_neighborhood_type`
  - [ ] Also for the [`kl_div`](https://github.com/ceciledebezenac/segregation_index/blob/master/divergence/neighbourhood.py#L86), which might use `scipy.special.kl_div`for nan filtering & vectorized computation
- [ ] **Use `pysal` where possible**
  - [ ] [Adjacency computations](https://github.com/ceciledebezenac/segregation_index/blob/master/divergence/neighbourhood.py#L121) definitely would become a `pysal.weights.Rook` if when we integrate, or hit the `geodata.sindex` at least?
- [ ] smaller stuff
  - [ ] [print](https://github.com/ceciledebezenac/segregation_index/blob/master/divergence/segregation_distortion.py#L290) generally should use `warnings.warn` when showing warnings. 
  - [ ] [Maybe we should pull graphics imports](https://github.com/ceciledebezenac/segregation_index/blob/master/divergence/segregation_distortion.py#L8) inside of the plot methods & try/catch to avoid crashing? Most folks have these installed I think, but it's technically not necessary for the statistical functions, so that might be nice to avoid a hard dependency there? 
  - [ ] Unittests would be needed
  - [ ] documentation/examples should be made [generic](https://github.com/ceciledebezenac/segregation_index/blob/master/example_distortion.ipynb), so the data would have to be published or examples changed.

**Comment by knaaptime** [comment]
agree with Levi's points. also,

1) is pickle really required? if so, what for?
2) we have [gini](https://github.com/pysal/segregation/blob/83311ca71a957d3c32be96e798d677a525174fa4/segregation/aspatial/aspatial_indexes.py#L380) and [lorenz](https://github.com/pysal/segregation/blob/83311ca71a957d3c32be96e798d677a525174fa4/segregation/util/plot.py#L41) implementations (but we might prefer your lorenz implementation) so we might want to think about how reconfigure these pieces to fit together without duplication

**Comment by knaaptime** [comment]
matplotlib is already a dependency, but might be good to move seaborn into the function

**Comment by ceciledebezenac** [comment]
Hello all,
Thank you for going through the code so quickly! I totally agree with all that has been said, I paid very little attention to performance for computing the neighbourhood structure... I used pickle for my own analysis, but left it in the cleaner version for no particular reason.
The scipy.special.kl_div method is great and would make the code a lot simpler. 
Please let me know if there are other parts of it that are obscure...

**Comment by ljwolf** [comment]
Happy to help get this done :) this would be an awesome addition to the library!

**Comment by knaaptime** [comment]
also i just realized seaborn is a dependency elsewhere in the pysal stack (spvcm) so i'd be fine including it as a hard dependency here as well
